lib/src/Base/Optim/Pagmo.cxx

```diff
@@ -0,0 +1,627 @@
+  if (problem.hasResidualFunction() || problem.hasLevelFunction())
+    throw InvalidArgumentException(HERE) << "Pagmo does not support least squares or nearest point problems";
+  const Description multiObjectiveAgorithms = {"nsga2", "moead", "mhaco", "nspso"};
+  if ((problem.getObjective().getOutputDimension() > 1) && !multiObjectiveAgorithms.contains(getAlgorithmName()))
+    throw InvalidArgumentException(HERE) << getAlgorithmName() << " does not support multi-objective optimization";
+  if ((problem.getObjective().getOutputDimension() < 2) && multiObjectiveAgorithms.contains(getAlgorithmName()))
+    throw InvalidArgumentException(HERE) << getAlgorithmName() << " only supports multi-objective optimization";
+  const Description integerAgorithms = {"gaco", "ihs", "sga", "nsga2", "mhaco"};
+  if (!problem.isContinuous() && !integerAgorithms.contains(getAlgorithmName()))
+    throw InvalidArgumentException(HERE) << getAlgorithmName() << " does not support non continuous problems";
+#else
+  (void)problem;
+  throw InvalidArgumentException(HERE) << "No Pagmo support";
+#endif
+}
+
+
+void Pagmo::run()
+{
+  if (!startingSample_.getSize())
+    throw InvalidArgumentException(HERE) << "Starting sample is empty";
+  if (startingSample_.getDimension() != getProblem().getDimension())
+    throw InvalidArgumentException(HERE) << "Starting sample dimension does not match problem dimension";
+  for (UnsignedInteger i = 0; i < startingSample_.getSize(); ++ i)
+  {
+    const Point inP(startingSample_[i]);
+    if (!getProblem().getBounds().contains(inP))
+      LOGWARN(OSS() << "Starting point " << i << " lies outside bounds");
+    if (!getProblem().isContinuous())
+    {
+      const Indices types(getProblem().getVariablesType());
+      for (UnsignedInteger j = 0; j < getProblem().getDimension(); ++ j)
+      {
+        if (types[j] != OptimizationProblemImplementation::CONTINUOUS && std::trunc(inP[j]) != inP[j])
+          throw InvalidArgumentException(HERE) << "Starting sample components must have integral values";
+        if (types[j] == OptimizationProblemImplementation::BINARY && (std::round(inP[j]) != 0.0) && (std::round(inP[j]) != 1.0))
+          throw InvalidArgumentException(HERE) << "Starting sample components must have binary values";
+      }
+    }
+  }
+
+#ifdef OPENTURNS_HAVE_PAGMO
+  PagmoProblem pproblem(getProblem(), blockSize_);
+  pagmo::problem prob(pproblem);
+  const Description constrainedAgorithms = {"gaco", "ihs"};
+  Bool emulatedConstraints = false;
+  if ((getProblem().hasInequalityConstraint() || getProblem().hasEqualityConstraint()) && !constrainedAgorithms.contains(getAlgorithmName()))
+  {
+    emulatedConstraints = true;
+    prob = pagmo::unconstrain(prob, unconstrainMethod_);
+  }
+  pagmo::population pop(prob, 0, 0);
+
+  // nsga2 needs the population size to be a multiple of 4
+  const UnsignedInteger populationSize = (algoName_ == "nsga2") ? 4 * (startingSample_.getSize() / 4) : startingSample_.getSize();
```

Comment:
worse-case scenario we loose 3 points that does not seem too awful to raise an exception

Reply:
@jschueller You may have a look at https://github.com/openturns/openturns/blob/master/lib/src/Uncertainty/Distribution/EmpiricalBernsteinCopula.cxx#L163 to see how I managed a similar issue in EmpiricalBernsteinCopula.

Reply:
@mbaudin47 It is a replacement by the greatest multiple of 4 less or equal to `startingSample_`, a kind of `floor`, instead of a rounding. My english is too poor to find the correct wording for that, but rounding is misleading IMO.